### PR TITLE
220118 dhcp sethostname

### PR DESCRIPTION
**Pull request by pillip8282**

### CODE REVIEW

**Review by jeongchanKim — COMMENTED** [review]
Please check the indentation.
We recommend to use tab for indentation.

File: os/net/netmgr/netmgr_ioctl_lwip.c [hunk 1]
```diff
@@ -409,6 +425,13 @@ static int lwip_func_ioctl(int s, int cmd, void *arg)
 		}
 		ret = OK;
 		break;
+  case DHCPCSETHOSTNAME:
+    req->req_res = _netdev_dhcpc_sethostname(&req->msg.dhcp);
+    if (req->req_res != OK) {
+      NET_LOGKE(TAG, "Set dhcp host name failed %d\n", req->req_res);
+			goto errout;
+    }
+    ret = OK;
```
Comment (jeongchanKim) [inline os/net/netmgr/netmgr_ioctl_lwip.c:434]: There is no `break` in this case. 
Is it okay?

File: os/net/lwip/src/include/lwip/dhcp.h [hunk 2]
```diff
@@ -152,6 +157,15 @@ void dhcp_stop(struct netif *netif);
 
 #if LWIP_DHCP_HOSTNAME
 void dhcp_hostname(struct netif *netif, char *name);
+/**
+ * @brief dhcp_sethostname() set hotsname of dhcpc
+ *
+ * @internal
+ * @param[in] msg contains netif and hostname
```
Comment (jeongchanKim) [inline os/net/lwip/src/include/lwip/dhcp.h:164]: Please check the comment, there are two params, `netif` and `arg`.

File: external/include/protocols/dhcpc.h [hunk 3]
```diff
@@ -107,6 +107,18 @@ int dhcp_client_start(const char *intf);
  */
 void dhcp_client_stop(const char *intf);
 
+/**
+ * @brief Set hostname
+ *
+ * @param[in] intf name of interface to stop dhcpc
+ * @param[in] hostname hostname of interface
```
Comment (jeongchanKim) [inline external/include/protocols/dhcpc.h:114]: The 2nd param is `hostnameq`, but the comment says `hostname`.
Please check it.

File: os/net/lwip/src/include/lwip/dhcp.h (hunk 2 above)
Comment (pillip8282) [reply os/net/lwip/src/include/lwip/dhcp.h:164]: Thanks I'll modify it

File: external/include/protocols/dhcpc.h (hunk 3 above)
Comment (pillip8282) [reply external/include/protocols/dhcpc.h:114]: Thanks I'll modify it

File: apps/system/utils/netcmd.c [hunk 4]
```diff
@@ -240,6 +253,10 @@ static int _cmd_ifconfig_parse(int argc, char **argv, struct ifconfig_cmd_info_s
 
 	info->intf = argv[1];
 
+  if (strncmp(argv[2], "dhcp", strlen("dhcp") + 1) == 0) {
```
Comment (jeongchanKim) [inline apps/system/utils/netcmd.c:256]: Please check the indentation here also :)

File: external/dhcpc/dhcpc_lwip.c [hunk 5]
```diff
@@ -139,3 +139,30 @@ void dhcp_client_stop(const char *intf)
...
 }
+
+int dhcp_client_sethostname(const char *intf, const char *hostname)
+{
+	int ret = -1;
+	struct req_lwip_data req;
+
+	int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (sockfd < 0) {
+		printf("socket() failed with errno: %d\n", errno);
```
[the remark this answers is not in the document]
Comment (pillip8282) [reply external/dhcpc/dhcpc_lwip.c:150]: it's good opinion. I'll apply it.

File: os/include/tinyara/netmgr/netctl.h [hunk 6]
```diff
@@ -29,6 +29,7 @@ typedef enum {
 	DNSSETSERVER,
 	DHCPCSTART,
 	DHCPCSTOP,
+  DHCPCSETHOSTNAME,
```
Comment (jeongchanKim) [inline os/include/tinyara/netmgr/netctl.h:32]: Please check the indentation.

File: os/net/lwip/src/api/netifapi.c [hunk 7]
```diff
@@ -209,4 +227,25 @@ err_t netifapi_netif_common(struct netif *netif, netifapi_void_fn voidfunc, neti
...
+err_t netifapi_netif_common_arg(struct netif *netif, netifapi_void_fn_arg voidfunc, netifapi_errt_fn_arg errtfunc, void *arg)
+{
+	err_t err;
+	NETIFAPI_VAR_DECLARE(msg);
+	NETIFAPI_VAR_ALLOC(msg);
+
+	NETIFAPI_VAR_REF(msg).netif = netif;
+	NETIFAPI_VAR_REF(msg).msg.common_arg.voidfunc = voidfunc;
+	NETIFAPI_VAR_REF(msg).msg.common_arg.errtfunc = errtfunc;
+  NETIFAPI_VAR_REF(msg).msg.common_arg.arg = arg;
```
Comment (jeongchanKim) [inline os/net/lwip/src/api/netifapi.c:245]: Please check the indentation.

File: os/net/netmgr/netmgr_ioctl_lwip.c [hunk 8]
```diff
@@ -409,6 +431,14 @@ static int lwip_func_ioctl(int s, int cmd, void *arg)
...
 		ret = OK;
 		break;
+	case DHCPCSETHOSTNAME:
+		req->req_res = _netdev_dhcpc_sethostname(&req->msg.dhcp);
+		if (req->req_res != OK) {
+			NET_LOGKE(TAG, "Set dhcp host name failed %d\n", req->req_res);
+			goto errout;
+		}
+		ret = OK;
+    break;
```
Comment (jeongchanKim) [inline os/net/netmgr/netmgr_ioctl_lwip.c:441]: Please check the indentation.

File: os/include/tinyara/netmgr/netctl.h (hunk 6 above)
Comment (pillip8282) [reply os/include/tinyara/netmgr/netctl.h]: Thanks for review. I'll modify it

File: os/net/lwip/src/include/lwip/netifapi.h [hunk 9]
```diff
@@ -85,6 +87,11 @@ struct netifapi_msg {
 			netifapi_void_fn voidfunc;
 			netifapi_errt_fn errtfunc;
 		} common;
+    struct {
+      netifapi_void_fn_arg voidfunc;
+			netifapi_errt_fn_arg errtfunc;
```
Comment (sangam-swami) [inline os/net/lwip/src/include/lwip/netifapi.h:92]: can we do indentation for this structure

File: os/net/netmgr/netmgr_ioctl_lwip.c [hunk 10]
```diff
@@ -182,6 +182,28 @@ int _netdev_dhcpc_stop(const char *intf)
...
+    NET_LOGKE(TAG, "No network interface for dhcpc %s\n", msg->intf);
+    return ERROR;
+  }
+  dmsg.hostname = msg->hostname;
+
+  netifapi_dhcp_sethostname(dmsg.netif, (void *)&dmsg);
+
+  return 0;
+#else
+  NET_LOGKE(TAG, "Enable LWIP_DHCP_HOSNAME\n");
```
Comment (sangam-swami) [inline os/net/netmgr/netmgr_ioctl_lwip.c:203]: HOSTNAME

File: os/net/netmgr/netmgr_ioctl_lwip.c [hunk 11]
```diff
@@ -182,6 +182,28 @@ int _netdev_dhcpc_stop(const char *intf)
 
 	return OK;
 }
+
+
+int _netdev_dhcpc_sethostname(struct lwip_dhcp_msg *msg)
+{
+#ifdef CONFIG_LWIP_DHCP_HOSTNAME
+  // it's blocked call. So it's ok to use local variable
```
Comment (sangam-swami) [inline os/net/netmgr/netmgr_ioctl_lwip.c:190]: will use /* for comment lines?

Comment (pillip8282) [reply os/net/netmgr/netmgr_ioctl_lwip.c:190]: There is no coding rule about it. So It seems ok.

File: os/net/lwip/src/include/lwip/netifapi.h (hunk 9 above)
Comment (pillip8282) [reply os/net/lwip/src/include/lwip/netifapi.h:92]: Thanks I'll modify it

File: os/net/netmgr/netmgr_ioctl_lwip.c (hunk 10 above)
Comment (pillip8282) [reply os/net/netmgr/netmgr_ioctl_lwip.c]: Thanks I'll modify it

**Review by an4967 — APPROVED** [review]
LGTM

**Pull request merged**